<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>fwdpp: Example programs</title>
<link href="../../tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../jquery.js"></script>
<script type="text/javascript" src="../../dynsections.js"></script>
<link href="../../search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="../../doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">fwdpp
   &#160;<span id="projectnumber">0.2.9</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "../../search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="../../index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="../../pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="../../modules.html"><span>Modules</span></a></li>
      <li><a href="../../namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="../../annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="../../files.html"><span>Files</span></a></li>
      <li><a href="../../examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="../../search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="../../search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Data Structures</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">Example programs </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>The best documentation of how to use the library are the example simulations in the examples/ subdirectory.</p>
<h3>Compiling the examples:</h3>
<p>As of fwdpp 0.2.5, the examples are compiled by saying "make check" after the configure step.</p>
<h3>Running the examples:</h3>
<p>This following list of parameters and their definitions is in common to all of the example programs:<br/>
 N = the number of diploid individuals to simulate<br/>
 theta = 4Nu, where u is the mutation rate per gamete per generation. For example, theta = 100 means that on average, 2Nu = 50 new mutations are entering the population each generation.<br/>
 rho = 4Nr, where r is the recombination rate per diploid per generation<br/>
 g = the number of generations to simulate. Often, this should be &gt;= 8N at a minimum<br/>
 n = the sample size to draw at the end of the simulation. To match the typical modeling assumptions of population genetics, you should have n &lt;&lt; 2N.<br/>
 nreps = the number of replicates to simulate<br/>
 seed = a random number seed. I will use $RANDOM as a seed, referring to the bash shell method to return a random integer.<br/>
</p>
<p>Note: familiarity with Hudson's "[ms](http://home.uchicago.edu/~rhudson1)" program is helpful for some of what comes below.</p>
<h4>diploid</h4>
<p>The first program, diploid, simulates a Wright-Fisher population with mutation, recombination, and drift. To run it:</p>
<div class="fragment"><div class="line">./diploid N theta rho g n nreps seed | gzip &gt; outfile.gz</div>
</div><!-- fragment --><p>The data in outfile.gz will be in the same format as Dick Hudson's "ms" coalescent simulator.</p>
<p>Example:</p>
<div class="fragment"><div class="line">./diploid 10000 10 10 100000 50 1 $RANDOM | gzip &gt; test_diploid.gz</div>
</div><!-- fragment --><h4>diploid_ind</h4>
<p>Identical to diploids, but uses the individual-based sampler. As this program (and diploid) simulate neutral models, this program will typically be slower than that gamete-based implementation when N is large.</p>
<h4>diploid_binaryIO</h4>
<p>The next program is called diploid_binaryIO. This program is identical to diploid, except that it only simulates one replicate at a time, creates two output files, and outputs the entire population rather than a sample of size n &lt;&lt; 2N. The first file is an index file, containing an integer representing the replicate number of the output and the position in the haplotypes file where this record begins. The second file is the haplotypes file, which contains the entire population in binary format.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./diploid_binaryIO N theta rho g replicate\_number index\_filename haplotype\_filename seed.</div>
</div><!-- fragment --><p>Example for Open Grid Engine compute clusters:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#!sh</span></div>
<div class="line"><span class="preprocessor"></span><span class="preprocessor">#$ -t 1-100</span></div>
<div class="line"><span class="preprocessor"></span><span class="preprocessor">#$ -N DIPBIN</span></div>
<div class="line"><span class="preprocessor"></span></div>
<div class="line">seed=`echo <span class="stringliteral">&quot;$SGE_TASK_ID*$RANDOM&quot;</span>|bc -l`</div>
</div><!-- fragment --><p>#note: the below assumes that the binary is in the users's $PATH and that the GE system knows how to link it to the relevant dynamic libraries</p>
<div class="fragment"><div class="line">diploid_binaryIO 10000 10 10 100000 $SGE_TASK_ID indexfile hapfile $seed</div>
</div><!-- fragment --><p>The above script, when submitted to a Grid Engine queue, will result in 100 populations of size N=10,000 being written to hapfile. Further, “indexfile” will contain the ID number and position of each file. Records are not over-written because the program uses POSIX file locking to ensure that only 1 process at a time can do the writing. This is a complex program, as it mixes C++ objects with output streams such that they can be written to C-style file descriptors, which is required in order to use file locking (which is a C feature with no C++ analog). However, the advantage is that you write all data to one large file, avoiding the plague of lots of small files that can bring distributed file systems to their knees.</p>
<h4>diploid_binaryIO_ind</h4>
<p>Identical to diploid_binaryIO, but individual-based.</p>
<h4>diploid_fixed_sh</h4>
<p>This program is similar to diploid, but adds an additional mutation rate (theta_selected = 4Nu_s, where u_s is the mutation rate per gamete per generation to mutations with selection coefficient s) to mutations with selection coefficient s and dominance h. Fitness across sites is multiplicative. The output is in "ms" format&ndash;one block for neutral mutations followed by one block for selected mutations.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./dipoid_fixed_sh N theta theta_selected rho s h g n nreps seed | gzip &gt; outfile.gz</div>
</div><!-- fragment --><p>For this program, s can be positive or negative, as can h.</p>
<h4>diploid_fixed_sh_ind and diploid_fixed_sh_lambda</h4>
<p>Identical to diploid_fixed_sh, but based on the individual-based sampler.</p>
<p>These two programs are identical in function, but differ in that the former is implemented using std::bind and the latter using lambda expressions.</p>
<h4>diploid_twopop_mig</h4>
<p>This program simulates an ancestral population for g generations, at which point a daughter population of size N is “budded” off from the ancestral population. Evolution continues for g2 more generations, with symmetric migration at rate M = 4Nm, where m is the migration rate per diploid per generation. The output is in "ms" format, with n haplotypes per sample just like how ms outputs data for multi-population models.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./diploid_twopop_mig N theta rho g g2 M n nreps seed | gzip &gt; outfile.gz</div>
</div><!-- fragment --><p>Note: the demographic model here implemented may be viewed as biologically bizarre, as it mimics the default behavior of “ms” for population split models. Let’s compare a specific example vs. the equivalent ms command line:</p>
<div class="fragment"><div class="line">./diplod_twopop_mig 10000 50 50 100000 1000 1 50 1 $RANDOM | gzip &gt; outfile.gz</div>
</div><!-- fragment --><p>and</p>
<div class="fragment"><div class="line">ms 100 1 -t 50 -r 50 1000 -I 2 2 1 -ej 0.025 2 1 -em 0.025 1 2 0.</div>
</div><!-- fragment --><p>Why may this be considered odd? In ms, when two populations are merged, the rate of coalescence is unaffected by default (this behavior is documented, and it is up to the user to adjust population sizes when population merge and split in ms). That means when the two populations, each of size N merge, the merged (ancestral) population is still of size N. diploid_twopop_mig is doing the same thing forwards in time: an ancestral population of size N magically changes into two populations of size N.</p>
<h4>migsel_ind</h4>
<p>Simulates 2 equal-sized populations of size N (N remains constant over time) diploids with selection at strength s and dominance h. Migration occurs between the two populations.</p>
<p><em>Note that the two populations never share a common ancestor in this simulation! In other words, deme 1 is never founded by a sampling event from deme 0. This means that if you simulate for a short period of time, it is unlikely that there is a MRCA common to all individuals sampled. (See comment above about appropriateness of these examples for research...)</em></p>
<p>Usage:</p>
<div class="fragment"><div class="line">./migsel_ind N 4Nu_neut 4Nu_sel 4Nr 4Nm s h f1 f2 ngens n outfilename seed</div>
</div><!-- fragment --><p>where: N = population number for each deme. 4Nu_neut = 4N times the neutral mutation rate per gamete 4Nu_sel = 4N times the mutation rate per gamete to selected mutations 4Nr = 4N times the recombination rate per diploid per region per generation 4Nm = 4N times the probability of migration per diploid per generation s = the selection coefficient of newly-arising selected mutations h = dominance of newly-arising selected mutations f1 and f2 are the probabilities of inbreeding in deme 1 and 2, respectively ngens = # of generations to simulate n = sample size to take from each deme @ end of simulation (must be even #) outfilename = name of output file seed = random number seed.</p>
<p>Notes: s is taken to be s in deme 1 and -s in deme 2. This lets me illustrate how different fitness functions can be passed to different demes using fwdpp.</p>
<p>The output file contains the following:</p>
<ol type="1">
<li>The metapopulation in binary format</li>
<li>Two "ms"-format blocks of size 2*n. These are for neutral and selected mutations, respectively. Within each block, the first n haplotypes are from deme 1 and the second n are from deme 2.</li>
</ol>
<p>The program writes the data to the output file and then reads it in again. This is mainly to illustrate the binary I/O routines for individual-based metapopulation simulations.</p>
<h4>migsel_split_ind and migsel_split_ind_list</h4>
<p>Simulates a constant-sized population of N diploids with selection at strength s and dominance h for ngens generations. Then, a copy is made of the population, and the two demes are simulated with migration for another ngens2 generations.</p>
<p>These two programs also serialize their data, read the copied data back, and make sure that input equals output.</p>
<p>These two examples are hints on how to code up "IM"-like models.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./migsel_split_ind N 4Nu_neut 4Nu_sel 4Nr 4Nm s h f1 f2 ngens ngens2 n seed</div>
</div><!-- fragment --><div class="fragment"><div class="line">./migsel_split_ind_list N 4Nu_neut 4Nu_sel 4Nr 4Nm s h f1 f2 ngens ngens2 n seed</div>
</div><!-- fragment --><p>where: N = population number for each deme. 4Nu_neut = 4N times the neutral mutation rate per gamete 4Nu_sel = 4N times the mutation rate per gamete to selected mutations 4Nr = 4N times the recombination rate per diploid per region per generation 4Nm = 4N times the probability of migration per diploid per generation s = the selection coefficient of newly-arising selected mutations h = dominance of newly-arising selected mutations f1 and f2 are the probabilities of inbreeding in deme 1 and 2, respectively ngens = # of generations to simulate ngens2 = # of generations to simulate after the split n = sample size to take from each deme @ end of simulation (must be even #) seed = random number seed.</p>
<p>Notes: s is taken to be s in deme 1 and -s in deme 2. This lets me illustrate how different fitness functions can be passed to different demes using fwdpp.</p>
<h4>RHH</h4>
<p>This program simulates the "recurrent hitch-hiking" model in which the expected number of fixations of beneficial codominant mutations per site per 4N generations is Lambda. The output is ms-format (just one block for neutral mutations). This model is one of the most thoroughly-studied models of the effect of selection on linked variation. See, for example:</p>
<ol type="1">
<li>Wiehe, T. H. and W. Stephan, 1993 Analysis of a genetic hitchhiking model, and its application to DNA polymorphism data from Drosophila melanogaster. Molecular Biology and Evolution 10: 842–854.</li>
<li>Kaplan, N. L., R. R. Hudson and C. H. Langley, 1989 The "hitchhiking effect" revisited. Genetics 123: 887–899.</li>
<li>Braverman, J. M., R. R. Hudson, N. L. Kaplan, C. H. Langley and W. Stephan, 1995 The hitchhiking effect on the site frequency spectrum of DNA polymorphisms. Genetics 140: 783–796.</li>
</ol>
<ol type="1">
<li>Przeworski, M., 2002 The Signature of Positive Selection at Randomly Chosen Loci. Genetics 160: 1179.</li>
</ol>
<p>The model in this program allows selected mutations within the region with neutral mutation rate theta (the so-called "neutral" or "sampled" region in the coalescent literature treating RHH via the structured coalescent). In general, the majority of selected sites will appear up to s/r_bp base pairs, where r is the recombination rate per base pair.</p>
<p>The population is simulated for g generations with only neutral mutations, followed by g2 more generations with neutral and selected mutations. The selection coefficient, s, must be &gt; 0, but the program doesn't check for this, so just enter an appropriate value please. The output is in ms format, and only contains data for neutral mutations, so any segregating selected mutations will not be output. The region with mutation rate theta is L base pairs long. L is really only used to calculate r_bp, which is rho/(4*N*(L-1)), as there are L-1 possible positions for recombination to occur in a region of L nucleotides.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./RHH N theta rho L s Lambda g1 g2 n nreps seed | gzip &gt; outfile.gz</div>
</div><!-- fragment --><p>Note: many of the well-known formulas for the effect of RHH on linked, neutral variation make very strong assumptions about the parameter values. For example, N needs to be large and s needs to be small, but Ns needs to be large. Further, Lambda needs to be sufficiently small such that sweeps are independent in time. This means that plugging in values to this program and comparing to theoretical predictions may lead to apparent discrepancies. This is also the case with various coalescent simulations of RHH.</p>
<h4>bneck_selection_ind</h4>
<p>This program simulates a population for g generations at size N. In generation g+1, N changes to N2 &lt;= N. The population then grows exponentially to size N3 &gt;= N2 in g2 generations. Selected and neutral mutations are allowed each generation. The output is in “ms” format&ndash;one block for neutral mutations followed by one block for selected mutations.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./bneck_selection_ind N theta_neutral theta_sel rho s h g1 N2 N3 g2 n nreps seed</div>
</div><!-- fragment --><p>Where:<br/>
 N = starting population size<br/>
 theta_neutral = 4N*(neutral mutation rate per gamete)<br/>
 theta_sel = 4N*(mutation rate per gamete to selected mutations)<br/>
 rho = 4Nr, where r is recombination rate per diploid per generation<br/>
 s = selection coefficient. Can be negative or positive.<br/>
 h = dominance of selected mutations<br/>
 N2 = size of bottlenecked population<br/>
 N3 = size of recovered population<br/>
 g2 = generations taken to go from size N2 to size N3<br/>
 n = sample size to take from the population.<br/>
 nreps = # replicates to simulate<br/>
 seed = random number seed.<br/>
</p>
<h4>TFL2013</h4>
<p>This is the simulation from Thornton, Foran, and Long (2013) Properties and Modeling of GWAS when Complex Disease Risk Is Due to Non-Complementing, Deleterious Mutations in Genes of Large Effect. PLoS Genetics 9: e1003258. The manuscript is available <a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1003258">here</a>. See that paper for modeling details.</p>
<p>Usage:</p>
<div class="fragment"><div class="line">./TFL2013 N mu_disease mu_neutral littler esize dist sdE sdS burnin evolve msfile phenofile effectfile</div>
</div><!-- fragment --><p>Where:<br/>
 N = diploid population number<br/>
 mu_disease = mutation rate to causative mutations (per gamete)<br/>
 mu_neutral = mutation rate to neutral mutations (per gamete)<br/>
 littler = Recombination rate per diploid<br/>
 esize = Effect size of causative mutation.<br/>
 dist = boolean. If 0, then esize is fixed. If 1, then esize is the mean of an exponentural with mean esize<br/>
 sdE = standard deviation in Gaussian noise added to phenotype<br/>
 sdS = standard deviation in Gaussian fitness function. (Should probably just set to 1.)<br/>
 burnin = number of generations to evolve with no causative mutations.<br/>
 evolve = number of generations to evolve with neutral and causative mutations.<br/>
 msfile = File to write the entire population in "ms" format. Each sequential pair of gametes is a different diploids. Output is gzipped.<br/>
 phenofile = File to write phenotypes of population. Two columns: genetic component first, then Gaussian noise component. The phenotype is the sum of the two. Order of phenotypes is same as order of diploids in msfile. Gzipped<br/>
 effectfile = File to write positions and effect sizes of causative mutations. Gzipped.<br/>
</p>
<p>Example from the PLoS Genetics paper:</p>
<div class="fragment"><div class="line">./TFL2013 20000 0.000125 0.00125 0.00125 0.5 1 0.075 1 0 160000 msfile.gz phenotypes.gz effects.gz</div>
</div><!-- fragment --><p>will simulate a mean effect size of 0.5 (exponentially-distributed). The above will take probably many hours to run.</p>
<h4>diploid_ind_2locus</h4>
<p>This program simulates a 2-locus neutral model. Mutations arise at rate <img class="formulaInl" alt="$\theta$" src="../../form_4.png"/> at each locus, and each locus recombines at rate <img class="formulaInl" alt="$\rho$" src="../../form_5.png"/>. The recombination rate between loci is <img class="formulaInl" alt="$r_{bw}$" src="../../form_6.png"/>, which corresponds to the probability a crossover is observed between loci. In other words, the genotypes at the two loci will switch from </p>
<p class="formulaDsp">
<img class="formulaDsp" alt="\[\frac{AB}{ab}\]" src="../../form_7.png"/>
</p>
<p> to </p>
<p class="formulaDsp">
<img class="formulaDsp" alt="\[\frac{Ab}{aB}\]" src="../../form_8.png"/>
</p>
<p> with probability <img class="formulaInl" alt="$r_{bw}$" src="../../form_6.png"/>.</p>
<p>The output of the simulation is an "ms"-style block for each locus. Positions are uniform on the interval <img class="formulaInl" alt="$[0,1)$" src="../../form_9.png"/> at locus 1, and uniform on the interval <img class="formulaInl" alt="$[1,2)$" src="../../form_10.png"/> at locus 2.</p>
<p>The usage is:</p>
<div class="fragment"><div class="line">Usage:</div>
<div class="line">./diploid_ind_2locus N theta rho ngens n nreps seed</div>
<div class="line">Where:</div>
<div class="line">N = population size (number of diploids)</div>
<div class="line">theta = 4Nu, the scaled neutral mutation rate</div>
<div class="line">rho = 4Nr, the scale recombination rate</div>
<div class="line">rbw = the probability that the two loci cross over, per generation</div>
<div class="line">ngens = the number of generations to simulate</div>
<div class="line">n = the sample size to pull from the population at the end of each simulated replicate</div>
<div class="line">nreps = the number of replicates to simulated</div>
<div class="line">seed = seed value <span class="keywordflow">for</span> random number generations</div>
</div><!-- fragment --> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="../../doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
