<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>fwdpp: Tutorial 1: Policies in fwdpp</title>
<link href="../../tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../jquery.js"></script>
<script type="text/javascript" src="../../dynsections.js"></script>
<link href="../../search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="../../doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">fwdpp
   &#160;<span id="projectnumber">0.2.7</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "../../search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="../../index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="../../pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="../../modules.html"><span>Modules</span></a></li>
      <li><a href="../../namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="../../annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="../../files.html"><span>Files</span></a></li>
      <li><a href="../../examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="../../search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="../../search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Data Structures</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Macros</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Groups</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">Tutorial 1: Policies in fwdpp </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#TutIntro">Introduction</a></li>
<li class="level1"><a href="#TutGeneral">Policies in C++</a></li>
<li class="level1"><a href="#TutAlgo">Algorithms in fwdpp</a></li>
<li class="level1"><a href="#TutPolicies">Policy requirements in fwdpp</a><ul><li class="level2"><a href="#TutMut">Mutation policies</a></li>
<li class="level2"><a href="#TutRec">Recombination</a></li>
<li class="level2"><a href="#TutMig">Migration</a></li>
<li class="level2"><a href="#TutFitness">Fitness</a></li>
<li class="level2"><a href="#TutUpdateRemove">Updating and removal policies</a></li>
</ul>
</li>
<li class="level1"><a href="#TutTieItup">Putting it all togeter (kinda)</a></li>
</ul>
</div>
<div class="textblock"><h1><a class="anchor" id="TutIntro"></a>
Introduction</h1>
<p>This document is intended to be an in-depth discussion of policies and their role in implementing forward-time population genetic simulations using the C++ template library <b>fwdpp</b>. We will first describe what policies are using standard C++ examples, and then we will get into the harder stuff.</p>
<p>In this document, I am implictly assuming that the library is being used to implement what I refer to as a "single-locus" simulation, by which I mean the simulation of a contiguous genomic stretch to which we apply some arbitrary model of mutation, recombination, and fitness. It is possible to use the same machinery to implement a "multi-locus" model of non-contiguous genomic segments, each with different mutation models, recombination models, and arbitrary linkage relationships amongst them. However, such simulations are arguably easier to impement using the multilocus API provided by the library, and I refer the users to the <a class="el" href="../../de/dbc/md_md_multiloc.html">documentation</a> for that part of the library.</p>
<p>An understanding of C++ fundamentals, including containers, their iterators, and how they relate to the standard algorithms, is assumed knowledge here.</p>
<h1><a class="anchor" id="TutGeneral"></a>
Policies in C++</h1>
<h2>Policies are everywhere</h2>
<p>Policies are a part of every programming language. Generally-speaking, they modify the behavior of what functions are with (or to) data. In other words, they turn a generic function into a piece of code doing a specific task. Let's start with the rather trivial example of sorting a vector:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;algorithm&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;vector&gt;</span></div>
<div class="line"></div>
<div class="line"><span class="keyword">using namespace </span>std;</div>
<div class="line"></div>
<div class="line"><span class="keywordtype">int</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a3c04138a5bfe5d72780bb7e82a18e627">main</a> ( <span class="keywordtype">int</span> argc, <span class="keywordtype">char</span> ** argv )</div>
<div class="line">{</div>
<div class="line">  vector&lt;unsigned&gt; vu{10,9,8,7,6,5,4,3,2,1};</div>
<div class="line"></div>
<div class="line">  sort(vu.begin(),vu.end());</div>
<div class="line">}</div>
</div><!-- fragment --><p>The above example is 100% standard C++. But what is going on under the hood of the sort function is quite interesting. A sort algorithm is being executed, and the values are being compared via a call to this function:</p>
<div class="fragment"><div class="line"><span class="keyword">template</span> &lt;<span class="keyword">class</span> T&gt; <span class="keyword">struct </span>less : binary_function &lt;T,T,bool&gt; {</div>
<div class="line">  <span class="comment">//algorithms can&#39;t guess return types,</span></div>
<div class="line">  <span class="comment">//and therefore often need this typedef</span></div>
<div class="line">  <span class="keyword">typedef</span> <span class="keywordtype">bool</span> result_type;</div>
<div class="line">  <span class="keywordtype">bool</span> operator() (<span class="keyword">const</span> T&amp; x, <span class="keyword">const</span> T&amp; y)<span class="keyword"> const </span>{<span class="keywordflow">return</span> x&lt;y;}</div>
<div class="line">};</div>
</div><!-- fragment --><h2>Policies are often function objects</h2>
<p>The structure called less is a ``function object'', and is the policy being employed in the sort. Further, is a template, meaning it works on any data type for which <b>operator&lt;__ is defined. The detail that less inherits from __std::binary_function</b> is important for how it plugs into the sort algorithm, but we'll deal with those issues later.</p>
<p>The way such a function object is used looks like:</p>
<div class="fragment"><div class="line">      <span class="keywordtype">int</span> x = 5, y = 6;</div>
<div class="line"><span class="comment">/*</span></div>
<div class="line"><span class="comment">The less() instantiates a (temporary) object of type less&lt;T&gt;,</span></div>
<div class="line"><span class="comment">where T = int in this case.</span></div>
<div class="line"><span class="comment">the (x,y) passes those to variables to the operator() of less.</span></div>
<div class="line"><span class="comment">(This is where the term</span></div>
<div class="line"><span class="comment">``function object&#39;&#39;, or functor for short, comes from.)</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">bool x_is_less = less()(x,y);</span></div>
</div><!-- fragment --><h3>Policies change behavior of algorithms</h3>
<p>OK, so now we hopefully have a basic understanding of what a policy is and that algorithms in C++ work through policies implemented as function objects. This lets us change the behavior of algorithms:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;functional&gt;</span> <span class="comment">//need this header for std::greater&lt;T&gt;</span></div>
<div class="line"><span class="comment">//Sort in descending order (biggest values @ front of vu after sort)</span></div>
<div class="line">sort( vu.begin(), vu.end(), greater&lt;int&gt;() );</div>
</div><!-- fragment --><p>Same sort, different outcome because of different policy.</p>
<h3>Binding extends what policies can do</h3>
<p>We can further modify the behavior of policies by sending additional arguments along with the policy as it goes to the algorithm. This is called binding an argument to a function call. For example, let's find the first value in our vector that is &gt;= 5:</p>
<div class="fragment"><div class="line"><span class="comment">//we need these headers</span></div>
<div class="line"><span class="preprocessor">#include &lt;functional&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;iostream&gt;</span></div>
<div class="line"></div>
<div class="line">vector&lt;unsigned&gt;::iterator itr = find_if( vu.begin(), </div>
<div class="line">vu.end(), </div>
<div class="line">bind2nd( greater_equal&lt;int&gt;(), 5 ) );</div>
<div class="line"></div>
<div class="line"><span class="comment">//print out the value pointed to by the iterator, </span></div>
<div class="line"><span class="comment">//if and only if we found something.  In C++,</span></div>
<div class="line"><span class="comment">//when a policy never finds anything, </span></div>
<div class="line"><span class="comment">//the end of the container is returned by tradition</span></div>
<div class="line"><span class="keywordflow">if</span>( itr != vu.end() )</div>
<div class="line">{</div>
<div class="line">  cout &lt;&lt; *itr &lt;&lt; <span class="charliteral">&#39;\n&#39;</span>;</div>
<div class="line">}</div>
</div><!-- fragment --><p>The find_if algorithm takes each value in the range and evaluates it via the policy. Here, the policy is greater_equal, which takes two arguments. The second argument is provided by using the standard-library function bind2nd, which results in the value 5 being the second value passed to the policy's operator(). The new binders in C++11 are much better:</p>
<div class="fragment"><div class="line"><span class="comment">//we need these headers</span></div>
<div class="line"><span class="preprocessor">#include &lt;functional&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;iostream&gt;</span></div>
<div class="line"></div>
<div class="line">vector&lt;unsigned&gt;::iterator itr = find_if( vu.begin(), </div>
<div class="line">vu.end(), </div>
<div class="line"><span class="comment">//here, _1 is a placeholder for a value that the algorithm must provide</span></div>
<div class="line">std::bind( greater_equal&lt;int&gt;(),std::placeholders::_1, 5 ) );</div>
<div class="line"></div>
<div class="line"><span class="comment">//print out the value pointed to by the iterator, </span></div>
<div class="line"><span class="comment">//if and only if we found something.  In C++,</span></div>
<div class="line"><span class="comment">//when a policy never finds anything, </span></div>
<div class="line"><span class="comment">//the end of the container is returned by tradition</span></div>
<div class="line"><span class="keywordflow">if</span>( itr != vu.end() )</div>
<div class="line">{</div>
<div class="line">  cout &lt;&lt; *itr &lt;&lt; <span class="charliteral">&#39;\n&#39;</span>;</div>
<div class="line">}</div>
</div><!-- fragment --><h2>Summary so far</h2>
<ul>
<li>Policies change how algorithms behave</li>
<li>Policies are often templates</li>
<li>Policies are often function objects</li>
<li>Policies + binders + algorithms = a reusable code base that can do lots of different (and often quite complicated) things when the right policy is written.</li>
<li>Policies can often be quite short to implement (see the definition of less above). This doesn't have to be the case, but it often works out that way in practice.</li>
</ul>
<h1><a class="anchor" id="TutAlgo"></a>
Algorithms in fwdpp</h1>
<p>For individual-based simulations, the primary algorithms are <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a></b> and <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a86a4fcecb15c101816ccdd15eefd6a1b" title="Remove mutations from population Removes mutations that are lost. ">KTfwd::remove_lost</a></b> and/or <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a84e666ea27d38bdfc539ffd23653eaa9" title="Remove mutations from population Removes mutations that are fixed or lost. ">KTfwd::remove_fixed_lost</a></b>. For gamete-baesd simulations, additional algorithms include the functions <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a9566be0bd568e2e75b02b1f46457c25c" title="Apply mutation model to a population. ">KTfwd::mutate</a></b>, <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a41bccb4fd1108616d626e371c3cf83c4" title="Recombination Recombination. ">KTfwd::recombine</a></b>, and <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#ad1d1af58a12e56b980a03b2056770370">KTfwd::migrate</a></b> (and/or <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#afc5980168c14a3c1d55790e506a49336">KTfwd::migrate_from_to</a></b>). All of these functions are in the name space <b><a class="el" href="../../da/d8b/namespaceKTfwd.html" title="The primary namespace defined by this library. ">KTfwd</a></b> and are documented both in the source code, via the doxygen output based on the source code, and finally in the example code that comes with the library.</p>
<p>Some of the code in some of these algorithms is quite complex, largely because generic templates can have hideous syntax. You should't have to worry about that unless you like seeing how the sausage is made. Most users of <b>fwdpp</b> will be writing custom policies to stick into these algorithms and will likely be starting from the examples in order to get oriented. For them, the necessary information is to understand what is required of a policy. That is the subject of the remainder of this document.</p>
<h1><a class="anchor" id="TutPolicies"></a>
Policy requirements in fwdpp</h1>
<p>This section discusses the requirements placed on policies in <b>fwdpp</b>. These requirements are essentially standards placed on data types in order to ensure that simulations behave properly. (Note that <code>behave properly'' is not the same as</code>are implemented correctly''! It is totally possible to have a simulation that compiles with no warnings and runs without crashing but is not the model you had in mind.) The policy requirements are enforced during compilation, such that a nonconforming policy cannot result in a compiled simulation program.</p>
<p>In the following sections, we will discuss policy requirements, how the built-in policies work, and also create some new policies. We'll build some policies for a simple quantitative trait with selection simulation, and later on combine them all to see how <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a></b> would actually be called (for the case of a constant-sized population in an individual-based simulation). For additional and more complex examples, see the code for the example programs that come with the library.</p>
<p>Caveat emptor:</p>
<ol type="1">
<li>This document has been written ad lib, and there may be errors in the implementation of policies below. (The policies that are directly from the library have been copy/pasted, and so will be ok.)</li>
<li>The code blocks below are really pseudocode. The include directives are put in there as guides. Obviously, none of what is below constitutes the complete implementation of a simulation. For that, see the examples that come with the library.</li>
</ol>
<h2>Short version for the impatient</h2>
<p>In this section, I define the <em>minimal</em> requirements that a policy must conform to. I also provide pseudocode in the form of lambda expressions in order to illustrate these requirements. By <code>requirements'', I mean</code>things that must be true, otherwise a program using <b>fwdpp</b> will fail to compile.'' Because <b>fwdpp</b> is a template library, policy errors are caught at compile-time. (Biological errors in your policies, however, will result in run-time issues. For example, doing your fitness calculations incorrectly, etc.) When policies require more than the minimal requirements described here, the programmer sends them along via either the usual ``bind'' mechanisms or via capature in lambda expressions.</p>
<p>The example program <b>diploid_fixed_sh_lambda</b> is an example of writing all policies using lambda expressions&ndash;take a look there for a concrete example of these concepts. The lambda expressions are very nice in that they document the policy requirements completely, as one is required to write them in the lambda arguments.</p>
<p>For the sake of this example, let's assume that we are using the following types in an individual-based simulation (in C++11, ``using'' statements replace typedefs.):</p>
<div class="fragment"><div class="line"><span class="keyword">using</span> mtype = <a class="code" href="../../d1/d7f/structKTfwd_1_1mutation.html">KTfwd::mutation</a>;</div>
<div class="line"><span class="keyword">using</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#aeb0d6c95466be026b693c2c09289d58e">mlist</a> = std::list&lt;mtype&gt;;</div>
<div class="line"><span class="keyword">using</span> <a class="code" href="../../d6/d90/structKTfwd_1_1gamete__base.html">gtype</a> = <a class="code" href="../../d6/d90/structKTfwd_1_1gamete__base.html">KTfwd::gamete_base&lt;mtype&gt;</a>;</div>
<div class="line"><span class="keyword">using</span> <a class="code" href="../../d9/d1d/mlocusCrossoverTest_8cc.html#a43adc418ba4093dd1c7d3ad5dbf8fa3a">glist</a> = std::list&lt;gtype&gt;;</div>
</div><!-- fragment --><p>A minimal mutation policy takes a non-const pointer to an <b>mlist</b> as an argument, and returns an <b>mtype</b>:</p>
<div class="fragment"><div class="line"><span class="comment">//the [&amp;] is used to capture any things that may be needed</span></div>
<div class="line"><span class="keyword">auto</span> mut_policy = [&amp;](<a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#aeb0d6c95466be026b693c2c09289d58e">mlist</a> * __mutations){ <span class="keywordflow">return</span> function_generating_new_type( arguments ); };</div>
</div><!-- fragment --><p>We then need to decide how to add a new mutation (the return value of the mutation policy) to the list of mutations. This ``mutation insertion policy'' takes a const reference to an <b>mtype</b> and a non-const pointer to an <b>mlist</b> as arguments, and returns an <b>mlist::iterator</b>. This is an example for the infinite-sites model. Under this model, we know that a new mutation is at a new position, therefore we simply insert it at the end of the mlist:</p>
<div class="fragment"><div class="line"><span class="keyword">auto</span> mut_insertion_policy = [](<span class="keyword">const</span> mtype &amp; m,<a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#aeb0d6c95466be026b693c2c09289d58e">mlist</a> * __mutations) { </div>
<div class="line"><span class="keywordflow">return</span> __mutations-&gt;insert(__mutations-&gt;end(),m); </div>
<div class="line">};</div>
</div><!-- fragment --><p>When a gamete is mutated, we need to decide how to insert it into the <b>glist</b>. This ``gamete insertion policy'' takes a const reference to an <b>gtype</b> and a non-const pointer to an <b>glist</b> as arguments, and returns an <b>glist::iterator</b>. Again, under the infinitely-many sites model, a gamete with a new mutation is guaranteed to be unique, so we can simply insert it:</p>
<div class="fragment"><div class="line"><span class="keyword">auto</span> gamete_insertion_policy = [](<span class="keyword">const</span> <a class="code" href="../../d6/d90/structKTfwd_1_1gamete__base.html">gtype</a> &amp; g,<a class="code" href="../../d9/d1d/mlocusCrossoverTest_8cc.html#a43adc418ba4093dd1c7d3ad5dbf8fa3a">glist</a> * __gametes) { </div>
<div class="line"><span class="keywordflow">return</span> __gametes-&gt;insert(__gametes-&gt;end(),g); </div>
<div class="line">};</div>
</div><!-- fragment --><p>A fitness policy takes two const references to <b>glist::const_iterator</b> as arguments and returns a double:</p>
<div class="fragment"><div class="line"><span class="keyword">auto</span> fitness_policy = [&amp;](<span class="keyword">const</span> glist::const_iterator &amp; __gamete1,</div>
<div class="line"><span class="keyword">const</span> glist::const_iterator &amp; __gamete2) {</div>
<div class="line">  <span class="comment">//No selection!</span></div>
<div class="line">  <span class="keywordflow">return</span> 1.;</div>
<div class="line">}</div>
</div><!-- fragment --><p>See the header file <b><a class="el" href="../../db/dd7/fitness__models_8hpp.html">fwdpp/fitness_models.hpp</a></b> for examples of fitness policies that I provide for ``standard'' cases.</p>
<p>A recombination policy takes two non-const references to <b>glist::iterator</b>, scrambles up those gametes appropriately, and returns an unsigned integer representing the number of crossovers, etc., between the two gametes. Typically, this would work via a call to <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a8a46b107991548a40821e132496c1693" title="recombination for individual-based forward simulations ">KTfwd::recombine_gametes</a></b>, or the convenience wrapper <b><a class="el" href="../../d1/d5b/structKTfwd_1_1genetics101.html">KTfwd::genetics101</a></b>. Here is an example of the former, where the genetic map is uniform on the interval <img class="formulaInl" alt="$(0,1]$" src="../../form_12.png"/>:</p>
<div class="fragment"><div class="line"><span class="comment">//These parameters will be captuered by reference ([&amp;] in the lambda </span></div>
<div class="line"><span class="comment">//expression below), because their state will be modified</span></div>
<div class="line">gsl_rng * <a class="code" href="../../d9/d1d/mlocusCrossoverTest_8cc.html#a53b3e4bb1bd287d2150b9c128e15a316">r</a>;</div>
<div class="line"><a class="code" href="../../d9/d1d/mlocusCrossoverTest_8cc.html#a43adc418ba4093dd1c7d3ad5dbf8fa3a">glist</a> gametes;</div>
<div class="line"><span class="comment">//littler is the recombination rate per diploid per generation</span></div>
<div class="line"><span class="keyword">auto</span> rec_policy =  [&amp;](glist::iterator &amp; g1,</div>
<div class="line">glist::iterator &amp; g2) { <span class="keywordflow">return</span> <a class="code" href="../../da/d8b/namespaceKTfwd.html#a8a46b107991548a40821e132496c1693">KTfwd::recombine_gametes</a>(r,littler,&amp;gametes,g1,g2,</div>
<div class="line">  <span class="comment">//This nested lambda is our genetic map: uniform on interval (0,1]</span></div>
<div class="line">  [&amp;](){<span class="keywordflow">return</span> gsl_rng_uniform(r);}); },</div>
</div><!-- fragment --><p>I'll document migration policy functions at a later date, sorry.</p>
<h2><a class="anchor" id="TutMut"></a>
Mutation policies</h2>
<p>This is the mutation base class provided by <b>fwdpp</b></p>
<div class="fragment"><div class="line">  <span class="keyword">struct </span><a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html">KTfwd::mutation_base</a></div>
<div class="line">  {<span class="comment"></span></div>
<div class="line"><span class="comment">    /// Mutation position</span></div>
<div class="line"><span class="comment"></span>    <span class="keyword">mutable</span> <span class="keywordtype">double</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#aed95181f38c8e17b19a7a9da4b373b03">pos</a>;<span class="comment"></span></div>
<div class="line"><span class="comment">    /// Count of mutation in the population</span></div>
<div class="line"><span class="comment"></span>    <span class="keywordtype">unsigned</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#a9df75224a5f00b97dc09e68b475a517a">n</a>;<span class="comment"></span></div>
<div class="line"><span class="comment">    /// Is the mutation neutral or not?</span></div>
<div class="line"><span class="comment"></span>    <span class="keywordtype">bool</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#acfb3b4c94516cbbc5c28a12e1ad575b3">neutral</a>;<span class="comment"></span></div>
<div class="line"><span class="comment">    /// Used internally (don&#39;t worry about it for now...)</span></div>
<div class="line"><span class="comment"></span>    <span class="keywordtype">bool</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#ad4da8f49535997d3adc3ac9f524adaa5">checked</a>;</div>
<div class="line">    <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#a73b18577e5d041ad1822663335477862">mutation_base</a>(<span class="keyword">const</span> <span class="keywordtype">double</span> &amp; position, </div>
<div class="line">    <span class="keyword">const</span> <span class="keywordtype">unsigned</span> &amp; count, <span class="keyword">const</span> <span class="keywordtype">bool</span> &amp; isneutral = <span class="keyword">true</span>)</div>
<div class="line">      : pos(position),n(count),neutral(isneutral),<a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#ad4da8f49535997d3adc3ac9f524adaa5">checked</a>(false)</div>
<div class="line">    {   </div>
<div class="line">    }</div>
<div class="line">    <span class="keyword">virtual</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html#aa3f3247f207e45bb11a8bf418fbdcb67">~mutation_base</a>(){}</div>
<div class="line">};</div>
</div><!-- fragment --><p>The above code defines a mutation as something with a position (stored as a double), a count (unsigned integer), a boolean declaring the mutation to be neutral or not, and another boolean called ``checked'' which is very important but should only be directly manipulated by internal library functions (unless you really geek out and see what the internals are doing. In that case&ndash;go nuts.)</p>
<p>The mutation base class is not sufficient for any interesting sorts of simulations. Rather, one must derive a class from it with more data types. The library provides a class called mutation, which is probably the standard type of mutation that a population geneticist would think of (this class is also in the library's namespace <a class="el" href="../../da/d8b/namespaceKTfwd.html" title="The primary namespace defined by this library. ">KTfwd</a>):</p>
<div class="fragment"><div class="line">  <span class="keyword">struct </span>mutation : <span class="keyword">public</span> mutation_base<span class="comment"></span></div>
<div class="line"><span class="comment">  //!The simplest mutation type, adding just a selection </span></div>
<div class="line"><span class="comment"></span>  <span class="comment">//coefficient and dominance to the interface</span></div>
<div class="line">  {<span class="comment"></span></div>
<div class="line"><span class="comment">    /// selection coefficient</span></div>
<div class="line"><span class="comment"></span>    <span class="keyword">mutable</span> <span class="keywordtype">double</span> s;<span class="comment"></span></div>
<div class="line"><span class="comment">    /// dominance coefficient</span></div>
<div class="line"><span class="comment"></span>    <span class="keyword">mutable</span> <span class="keywordtype">double</span> h;</div>
<div class="line">    mutation( <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; position, <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; sel_coeff,<span class="keyword">const</span> <span class="keywordtype">unsigned</span> &amp; count,</div>
<div class="line">              <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; dominance = 0.5) </div>
<div class="line">      : mutation_base(position,count,(sel_coeff==0)),s(sel_coeff),h(dominance)</div>
<div class="line">    {</div>
<div class="line">    }</div>
<div class="line">    <span class="keywordtype">bool</span> operator==(<span class="keyword">const</span> mutation &amp; rhs)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">    </span>{</div>
<div class="line">      <span class="keywordflow">return</span>( fabs(this-&gt;pos-rhs.pos) &lt;= std::numeric_limits&lt;double&gt;::epsilon() &amp;&amp;</div>
<div class="line">              this-&gt;s == rhs.s );</div>
<div class="line">    }</div>
<div class="line">};</div>
</div><!-- fragment --><p>What does a mutation policy (model) need to do? <b>The answer is that a single call to the mutation model function (or function object) must return a single instance of the simulation's mutation type with a count of 1.</b></p>
<h3>Example: the infinitely-many sites model of mutation</h3>
<p>This mutation model states that a new mutation occurs at a site not currently segregating in the population. This statement implies the following:</p>
<ol type="1">
<li>We need a method to rapidly choose mutation positions that don't currently exist in the (meta-)population.</li>
<li>Each gamete containing a new mutation is by definition a new gamete in the (meta-)population. If we did #1 correctly, then the newly-mutated gamete differs from all others in the population by at least 1 new mutation.</li>
</ol>
<p>We will now implement this mutation model for the mutation type ``mutation'' defined above. In order to add some complexity to our mutation model, we will make the additional modeling assumptions:</p>
<ol type="1">
<li>Mutation positions are continuous on the interval <img class="formulaInl" alt="$[0,1)$" src="../../form_9.png"/>.</li>
<li>Neutral mutations arise at rate <img class="formulaInl" alt="$\mu$" src="../../form_21.png"/> per gamete per generation</li>
<li>Selected mutations arise at rate <img class="formulaInl" alt="$\mu_s$" src="../../form_22.png"/> per gamete per generation.</li>
<li>The selection coefficient for a newly-arising mutation is exponentially-distributed with mean <img class="formulaInl" alt="$s_m$" src="../../form_23.png"/>. Further, half the time, selected mutations are deleterious ( <img class="formulaInl" alt="$s < 0$" src="../../form_24.png"/>). Otherwise, they are beneficial ( <img class="formulaInl" alt="$s > 0$" src="../../form_25.png"/>).</li>
<li>Dominance will be uniform from 0 to 2. (We'll be scaling fitness as <img class="formulaInl" alt="$1, 1+hs, 1+2s$" src="../../form_26.png"/> for genotypes AA, Aa, and aa, respectively.)</li>
</ol>
<p>From a programming point of view, we need a means to lookup all mutation positions currently segregating in the population. <b>fwdpp</b> provides support for lookup tables that conform to the behavior of the type <b>std::map</b>. While one could use a type like</p>
<div class="fragment"><div class="line">std::map&lt;double,bool&gt;</div>
</div><!-- fragment --><p>it is more efficient to use a hash table like</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;unordered_set&gt;</span> <span class="comment">//need this header</span></div>
<div class="line"><span class="keyword">typedef</span> std::unordered_set&lt;double,std::hash&lt;double&gt;,<a class="code" href="../../dc/d8e/structKTfwd_1_1equal__eps.html">KTfwd::equal_eps</a> &gt; <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a0731e2ed14d3858427a2552e4944b778">lookup_table_type</a>;</div>
</div><!-- fragment --><p>The above code creates a new data type called <b>lookup_table_type</b> that hashes doubles with the data type <b><a class="el" href="../../dc/d8e/structKTfwd_1_1equal__eps.html" title="Returns true if std::max(lhs,rhs)-std::min(lhs,rhs) &lt;= std::numeric_limits&lt;T&gt;::epsilon() ...">KTfwd::equal_eps</a></b> as its comparison operator. That comparison operation is provided by the library and looks like this:</p>
<div class="fragment"><div class="line"><span class="keyword">struct </span>equal_eps</div>
<div class="line">{</div>
<div class="line">  <span class="keyword">typedef</span> <span class="keywordtype">bool</span> result_type;</div>
<div class="line">  <span class="keyword">template</span>&lt;<span class="keyword">typename</span> T&gt;</div>
<div class="line">  <span class="keyword">inline</span> <span class="keywordtype">bool</span> operator()(<span class="keyword">const</span> T &amp; lhs, <span class="keyword">const</span> T &amp; rhs)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">  </span>{</div>
<div class="line">    <span class="keywordflow">return</span>( std::max(lhs,rhs)-std::min(lhs,rhs) &lt;= std::numeric_limits&lt;T&gt;::epsilon() );</div>
<div class="line">  }</div>
<div class="line">};</div>
</div><!-- fragment --><p>Note that you could provide your own equality comparison policy for the hashing table. This one would be excellent, and should be included int the library in future versions as it may be the most robust:</p>
<div class="fragment"><div class="line"><span class="keyword">struct </span>equality_comparison_strict</div>
<div class="line">{</div>
<div class="line">  <span class="keyword">typedef</span> <span class="keywordtype">bool</span> result_type;</div>
<div class="line">  <span class="keyword">template</span>&lt;<span class="keyword">typename</span> T&gt;</div>
<div class="line">  <span class="keyword">inline</span> <span class="keywordtype">bool</span> operator()(<span class="keyword">const</span> T &amp; lhs, <span class="keyword">const</span> T &amp; rhs)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">  </span>{</div>
<div class="line">    <span class="keywordflow">return</span>( !(lhs &gt; rhs) &amp;&amp; !(lhs &lt; rhs) );</div>
<div class="line">  }</div>
<div class="line">};</div>
</div><!-- fragment --><p>We can now completely define our mutation model as a function (we could do it as a function object, too). In this example, We assume that we are using the boost list type and boost's memory pool allocator:</p>
<div class="fragment"><div class="line"><span class="keyword">typedef</span> <a class="code" href="../../d1/d7f/structKTfwd_1_1mutation.html">KTfwd::mutation</a> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>;</div>
<div class="line"><span class="keyword">typedef</span> boost::pool_allocator&lt;mtype&gt; <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a8da95455b301321cb0a4a7064ed8fe3d">mut_allocator</a>;</div>
<div class="line"><span class="keyword">typedef</span> boost::container::list&lt;mtype,mut_allocator &gt; <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#aeb0d6c95466be026b693c2c09289d58e">mlist</a>;</div>
<div class="line">mtype mutmodel( gsl_rng * r, mlist * mutations,</div>
<div class="line">                <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mu_neutral,</div>
<div class="line">                <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mu_selected,</div>
<div class="line">                <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mean_s,</div>
<div class="line">                lookup_table_type * lookup )</div>
<div class="line">  {</div>
<div class="line">    <span class="comment">//get new mutation position</span></div>
<div class="line">    <span class="keywordtype">double</span> pos = gsl_rng_uniform(r);</div>
<div class="line">    <span class="comment">//this is very rapid lookup...</span></div>
<div class="line">    <span class="keywordflow">while</span>( lookup-&gt;find(pos) != lookup-&gt;end() )</div>
<div class="line">    {</div>
<div class="line">      pos = gsl_rng_uniform(r);</div>
<div class="line">    }</div>
<div class="line">    <span class="comment">//ok, we have new position, so put it in lookup table</span></div>
<div class="line">    lookup-&gt;insert(pos);</div>
<div class="line"></div>
<div class="line">    <span class="comment">//law of TTL prob</span></div>
<div class="line">    <span class="keywordtype">bool</span> neutral = (gsl_rng_uniform(r) &lt;= (mu_neutral)/(mu_neutral+mu_selected)) ? <span class="keyword">true</span> : <span class="keyword">false</span>;</div>
<div class="line"></div>
<div class="line">    <span class="comment">//return neutral mutation</span></div>
<div class="line">    <span class="keywordflow">if</span> ( neutral ) { <span class="keywordflow">return</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>(pos,0,1,0); }</div>
<div class="line"></div>
<div class="line">    <span class="comment">//get selection coefficient</span></div>
<div class="line">    <span class="keywordtype">double</span> s = gsl_ran_exponential(r,mean_s);</div>
<div class="line">    <span class="keywordflow">if</span>( gsl_ran_uniform(r) &lt;= 0.5 ) { s = -1.*s; }</div>
<div class="line"></div>
<div class="line">    <span class="comment">//the gsl_ran_flat call generates the dominance</span></div>
<div class="line">    <span class="keywordflow">return</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>(pos,s,1,gsl_ran_flat(r,0.,2.));</div>
<div class="line">  }</div>
</div><!-- fragment --><p>That is is&ndash;the mutation model is complete. We still need to deal with how mutations are entered into data structures representing the population, but we'll treat that later.</p>
<p>The mutation policy is passed to any of the various <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a></b> functions in the library like this:</p>
<div class="fragment"><div class="line">std::bind(mutmodel,r,&amp;mutations,</div>
<div class="line">mu_neutral, mu_selected,</div>
<div class="line">mean_s, &amp;lookup);</div>
</div><!-- fragment --><p>Or, using C++11 lambda expressions:</p>
<div class="fragment"><div class="line">[&amp;](mlist * __mutations){ <span class="keywordflow">return</span> mutmodel(r,__mutations,mu_neutral,</div>
<div class="line">  mu_selected,mean_s,&amp;lookup); }</div>
</div><!-- fragment --><p>Note that several of the data types passed to the model are non-const pointers. Therefore, it is very likely that the data pointed to will be modified my the mutation model!</p>
<h3>A model for quantitative trait simulations</h3>
<p>Let's define a model where a mutation affecting fitness does so via its effect size, <img class="formulaInl" alt="$e$" src="../../form_27.png"/>, which is Gaussian-distributed with mean zero and standard deviation <img class="formulaInl" alt="$\sigma_e$" src="../../form_28.png"/>.</p>
<p>We need a mutation class:</p>
<div class="fragment"><div class="line"> <span class="keyword">struct </span>mut_e : <span class="keyword">public</span> <a class="code" href="../../d1/da4/structKTfwd_1_1mutation__base.html">KTfwd::mutation_base</a></div>
<div class="line"> {</div>
<div class="line">   <span class="keywordtype">double</span> e;</div>
<div class="line">   mut_e( <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; __pos,</div>
<div class="line">          <span class="keyword">const</span> <span class="keywordtype">unsigned</span> &amp; __n,</div>
<div class="line">          <span class="keyword">const</span> <span class="keywordtype">bool</span> &amp; __neut,</div>
<div class="line">          <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; __e ) : KTfwd::mutation_base(__pos,__n,__neut),e(__e)</div>
<div class="line">          {</div>
<div class="line">          }</div>
<div class="line"> };</div>
<div class="line"></div>
<div class="line"><span class="keyword">typedef</span> mut_e <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>;</div>
</div><!-- fragment --><p>OK, our mutation model is going to be the following. It is infinitely-many sites with both neutral and non-neutral mutations:</p>
<div class="fragment"><div class="line"><span class="keyword">typedef</span> <a class="code" href="../../d1/d7f/structKTfwd_1_1mutation.html">KTfwd::mutation</a> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>;</div>
<div class="line"><span class="keyword">typedef</span> boost::pool_allocator&lt;mtype&gt; <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a8da95455b301321cb0a4a7064ed8fe3d">mut_allocator</a>;</div>
<div class="line"><span class="keyword">typedef</span> boost::container::list&lt;mtype,mut_allocator &gt; <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#aeb0d6c95466be026b693c2c09289d58e">mlist</a>;</div>
<div class="line">mtype mutmodel_Qtrait( gsl_rng * r, mlist * mutations,</div>
<div class="line">                       <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mu_neutral,</div>
<div class="line">                       <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mu_selected,</div>
<div class="line">                       <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; sigma_e,</div>
<div class="line">                       lookup_table_type * lookup )</div>
<div class="line"> {</div>
<div class="line">   <span class="comment">//get new mutation position</span></div>
<div class="line">   <span class="keywordtype">double</span> pos = gsl_rng_uniform(r);</div>
<div class="line">   <span class="comment">//this is very rapid lookup...</span></div>
<div class="line">   <span class="keywordflow">while</span>( lookup-&gt;find(pos) != lookup-&gt;end() )</div>
<div class="line">   {</div>
<div class="line">     pos = gsl_rng_uniform(r);</div>
<div class="line">   }</div>
<div class="line">   <span class="comment">//ok, we have new position, so put it in lookup table</span></div>
<div class="line">   lookup-&gt;insert(pos);</div>
<div class="line"></div>
<div class="line">   <span class="comment">//law of TTL prob</span></div>
<div class="line">   <span class="keywordtype">bool</span> neutral = (gsl_rng_uniform(r) &lt;= (mu_neutral)/(mu_neutral+mu_selected)) ? <span class="keyword">true</span> : <span class="keyword">false</span>;</div>
<div class="line"></div>
<div class="line">   <span class="comment">//return neutral mutation</span></div>
<div class="line">   <span class="keywordflow">if</span> ( neutral ) { <span class="keywordflow">return</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>(pos,1,<span class="keyword">true</span>,0); }</div>
<div class="line"></div>
<div class="line">   <span class="comment">//the gsl_ran_gaussian call determines the effect size</span></div>
<div class="line">   <span class="keywordflow">return</span> <a class="code" href="../../d8/d48/bneck__selection__dist_8cc.html#a08e697ef28ea008d75cc6de09f177dfa">mtype</a>(pos,1,gsl_ran_gaussian(r,sigma_e),<span class="keyword">false</span>);</div>
<div class="line"> }</div>
</div><!-- fragment --><h2><a class="anchor" id="TutRec"></a>
Recombination</h2>
<h3>How recombination is modeled in <b>fwdpp</b></h3>
<p>Currently, recombination is modeled as follows. In a diploid, there number of crossovers between gametes is Poisson distributed with mean <img class="formulaInl" alt="$r$" src="../../form_19.png"/>. There is no notion of interference in establishing crossover positions. (However, that can be done with a clever policy.)</p>
<p>Note: what I describe as <img class="formulaInl" alt="$r$" src="../../form_19.png"/> above is the usual definition in population genetics. However, gamete-based simulations (confusingly, I admit) use r/2, because the per-generation r is used to determine how many recombination events a gamete is involved in. In other words, it is treated as a "per-gamete'' parameter. There is more detail on this in the main library documentation, and you'll see in the example simulations that I divide <img class="formulaInl" alt="$\rho$" src="../../form_5.png"/> (the population-scaled rate of recombination) by <img class="formulaInl" alt="$8N$" src="../../form_29.png"/> to get ``littler'' rather than the more common <img class="formulaInl" alt="$4N$" src="../../form_30.png"/>. (Again, to be clear, <img class="formulaInl" alt="$r=\rho/4N$" src="../../form_31.png"/> for individual-based simulations.)</p>
<h3>Recombination map functions</h3>
<p>In <b>fwdpp</b>, a recombination map is a function or function object that returns a double and takes no additional arguments from the algorithm. The return value is the position of the crossing over event. The simplest recombination map is uniform. Here is how to implement such a map on the interval <img class="formulaInl" alt="$[0,1)$" src="../../form_9.png"/>:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;functional&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;gsl/gls_rng.h&gt;</span></div>
<div class="line"><span class="comment">//gsl_rng * r assumed to be initialized already...</span></div>
<div class="line">std::function&lt; double(void) &gt; recmap = std::bind( gsl_rng_uniform, r);</div>
</div><!-- fragment --><p>Let us write a recombination map function that models a strong hotspot of crossing over. The positions will still be on the interval <img class="formulaInl" alt="$[0,1)$" src="../../form_9.png"/>. A fraction <img class="formulaInl" alt="$p$" src="../../form_32.png"/> of the recombination events will come from a uniform distribution and <img class="formulaInl" alt="$1-p$" src="../../form_33.png"/> will come from a beta distribution with parameters <img class="formulaInl" alt="$a$" src="../../form_34.png"/> and <img class="formulaInl" alt="$b$" src="../../form_35.png"/>. The genetic map looks like this:</p>
<div class="fragment"><div class="line"><span class="keywordtype">double</span> mixture_map(gsl_rng * r, </div>
<div class="line">                   <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; p,</div>
<div class="line">                   <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; a,</div>
<div class="line">                   <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; b)</div>
<div class="line">    {</div>
<div class="line">      <span class="keywordflow">return</span> (gsl_rng_uniform(r) &lt;= p) ? gsl_rng_uniform(r) : gsl_ran_beta(r,a,b);</div>
<div class="line">    }</div>
<div class="line"><span class="comment">/*</span></div>
<div class="line"><span class="comment">   This is a hot hotspot.  say hist( c(runif(1e3),rbeta(9e3,100,100) ) ) in R </span></div>
<div class="line"><span class="comment">   to see density of crossover positions</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">std::function&lt; double(void) &gt; recmap = std::bind( mixture_map,r,0.1,100,100);</span></div>
</div><!-- fragment --><p>Either of the above code blocks results in a variable called recmap which is a function object representing a function call that takes no additional arguments and returns a double. The variable recmap can be passed to the algorithm as the recombination (sometimes called genetic in the library documentation) map policy.</p>
<p>Note that the above policies were implemented by <em>synthesiszing</em> a new function object type from a <b>std::bind</b> operation via the <b>std::function</b> template class. You may synthesize all of your policies into variables this way, but it is not required. However, the next subsection will reveal a case where it is required.</p>
<p>Note that gamete-based simulations only require this policy.</p>
<h3>Recombination model policy functions</h3>
<p>Individual-based simulations require a recombination model policy. This policy is a function/function object that takes two non-const references to iterators to gametes as arguments and returns an unsigned integer representing the number of crossovers between the two gametes. (Note that the return value will never been seen by a library user. It basically exists to help debug things deeper in the library if and when it comes to that.)</p>
<p>The library provides the policy <b><a class="el" href="../../d1/d5b/structKTfwd_1_1genetics101.html">KTfwd::genetics101</a></b> which implements the model of crossing over described above.</p>
<p>The policy looks like this:</p>
<div class="fragment"><div class="line"><span class="keyword">struct </span>genetics101</div>
<div class="line">{</div>
<div class="line">  <span class="keyword">typedef</span> <span class="keywordtype">unsigned</span> result_type;</div>
<div class="line">  <span class="keyword">template</span>&lt;<span class="keyword">typename</span> gamete_iterator_type,</div>
<div class="line">           <span class="keyword">typename</span> gamete_list_type_allocator,</div>
<div class="line">           <span class="keyword">template</span>&lt;<span class="keyword">typename</span>,<span class="keyword">typename</span>&gt; <span class="keyword">class </span>gamete_list_type,</div>
<div class="line">           <span class="keyword">typename</span> rec_pos_generator&gt;</div>
<div class="line">  <span class="keywordtype">unsigned</span> operator()( gamete_iterator_type &amp; g1,</div>
<div class="line">                       gamete_iterator_type &amp; g2,</div>
<div class="line">                       gamete_list_type&lt; <span class="keyword">typename</span> gamete_iterator_type::value_type,</div>
<div class="line">                                         gamete_list_type_allocator &gt; * gametes,</div>
<div class="line">                                         <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; littler,</div>
<div class="line">                                         gsl_rng * r,</div>
<div class="line">                                         <span class="keyword">const</span> rec_pos_generator &amp; rp)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">     </span>{</div>
<div class="line">       <span class="keyword">typedef</span> gamete_list_type&lt; <span class="keyword">typename</span> gamete_iterator_type::value_type, </div>
<div class="line">                                 gamete_list_type_allocator &gt; glist_t;</div>
<div class="line">       <span class="keywordtype">unsigned</span> NREC = 0;</div>
<div class="line">       <span class="keywordflow">if</span>( g1 != g2 &amp;&amp; gsl_rng_uniform(r) &lt;= 0.5 )</div>
<div class="line">       <span class="comment">//then a non-parental type is inherited from p1 and p1 has two different gametes                                    </span></div>
<div class="line">       {</div>
<div class="line">         NREC += <a class="code" href="../../da/d8b/namespaceKTfwd.html#a8a46b107991548a40821e132496c1693">recombine_gametes</a>(r,littler,gametes,g1,g2,rp,</div>
<div class="line">         std::bind(update_if_exists_insert&lt;typename gamete_iterator_type::value_type,glist_t&gt;,</div>
<div class="line">         std::placeholders::_1,gametes));</div>
<div class="line">       }</div>
<div class="line">       <span class="keywordflow">return</span> NREC;</div>
<div class="line">     }</div>
<div class="line">   };</div>
</div><!-- fragment --><p>The main thing a library user needs to focus on is the argument list for <b>operator()</b>. Specifically, it requires a variable of type <b>rec_pos_generator</b> which is stated in the documentation to be a recombination map policy. Thus, a recombination model is a policy that requires another policy. Further, <b>a recombination policy is passed non-const references to iterators to two gametes (g1 and g2 in the code above). Those iterators are updated in place. In the event of no recombination, g1 and g2 remain unchanged. Otherwise, g1 and g2 are modified to point to the new recombinants, which are new gametes inersted into the population at a count of 1.</b></p>
<p>We pass this recombination model in an individual-based simulation to <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a></b> like this:</p>
<div class="fragment"><div class="line"><span class="keyword">using</span> std::placeholders; <span class="comment">//_1,_2, etc.</span></div>
<div class="line">std::bind(<a class="code" href="../../d1/d5b/structKTfwd_1_1genetics101.html">KTfwd::genetics101</a>(),  <span class="comment">//the rec. model</span></div>
<div class="line">           _1,_2,                  <span class="comment">//placeholder for iterators to gametes</span></div>
<div class="line">           &amp;gametes,               <span class="comment">//pointer to gamete list</span></div>
<div class="line">           littler,                <span class="comment">//Avg. # of crossovers b/w two gametes per region per generation</span></div>
<div class="line">           r,                      <span class="comment">//a gsl_rng *</span></div>
<div class="line">           recmap)                 <span class="comment">//A genetic map policy like the one we made above</span></div>
</div><!-- fragment --><p>Note: if you want to write a new recombination policy, you probably want to proceed by modifying how it interacts with the the recombination map policy. For example, if a recombination at position <img class="formulaInl" alt="$x$" src="../../form_36.png"/> means that the next position must be <img class="formulaInl" alt="$\geq 1.5x$" src="../../form_37.png"/> (in some model of interference). Doing so requires making a custom version of the <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a8a46b107991548a40821e132496c1693" title="recombination for individual-based forward simulations ">KTfwd::recombine_gametes</a></b> function. If you read the code for that function, you will see where the recombination map policy is called. If you try to modify the code below that, then good luck to you. It isn't super-complicated, but tread with caution.</p>
<h2><a class="anchor" id="TutMig"></a>
Migration</h2>
<p>Migration policies are only used in individual-based simulations. For gamete-based simulations, you may write a migration function (replacing <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#ad1d1af58a12e56b980a03b2056770370">KTfwd::migrate</a></b>) that does what you need it to and is implemented in terms of <b>KTfwdd::migrate_from_to</b>.</p>
<p>For individual-based simulations involving a metapopulations, parent 1 comes from population <img class="formulaInl" alt="$i$" src="../../form_16.png"/> and may or not be a migrant. Parent two comes from population <img class="formulaInl" alt="$j$" src="../../form_38.png"/> and <img class="formulaInl" alt="$j = i$" src="../../form_39.png"/> in the case of no migration, otherwise <img class="formulaInl" alt="$j \neq i$" src="../../form_40.png"/>. Migration policies may be the trickiest to write effectively because spatial models of migration can be complicated. However, a migration policies requirements are simple. __A migration policy is a function or function object taking an argument if type size_t and returning a value of type size_t. The argument is the index of population <img class="formulaInl" alt="$i$" src="../../form_16.png"/>, and the return value is the index of population <img class="formulaInl" alt="$j$" src="../../form_38.png"/>.__</p>
<p>For example, let's assume two demes with migration rate <img class="formulaInl" alt="$m$" src="../../form_41.png"/>. Here, <img class="formulaInl" alt="$m$" src="../../form_41.png"/> is the probability that a parent is a migrant. This migration rate is equal between the two demes. Because we are in a C-like language, the values allowed for the <b>size_t</b> are <img class="formulaInl" alt="$0 \leq i \leq 1$" src="../../form_42.png"/>. The migration policy is thus defined as follows:</p>
<div class="fragment"><div class="line"><span class="keywordtype">size_t</span> <a class="code" href="../../df/dfa/migsel__ind_8cc.html#a4862499cb62bfa43943001b59fe6dc0d">migpop</a>(<span class="keyword">const</span> <span class="keywordtype">size_t</span> &amp; source_pop, gsl_rng * r, <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; mig_prob)</div>
<div class="line">{</div>
<div class="line">  <span class="comment">//if parent is a migrant</span></div>
<div class="line">  <span class="keywordflow">if</span>( gsl_rng_uniform(r) &lt;= mig_prob )</div>
<div class="line">  {</div>
<div class="line">    <span class="comment">//return other population</span></div>
<div class="line">    <span class="keywordflow">return</span> ! source_pop;</div>
<div class="line">  }</div>
<div class="line">  <span class="comment">//else, not a migrant</span></div>
<div class="line">  <span class="keywordflow">return</span> source_pop;</div>
<div class="line">}</div>
</div><!-- fragment --><p>And we pass it to <b>KTfwd::ample_diploid</b> like this:</p>
<div class="fragment"><div class="line">std::bind(<a class="code" href="../../df/dfa/migsel__ind_8cc.html#a4862499cb62bfa43943001b59fe6dc0d">migpop</a>, <span class="comment">//the policy</span></div>
<div class="line">std::placeholders::_1,     <span class="comment">//placeholder for population index i</span></div>
<div class="line">r,     <span class="comment">//gsl_rng *</span></div>
<div class="line">m)     <span class="comment">//migration rate</span></div>
</div><!-- fragment --><h2><a class="anchor" id="TutFitness"></a>
Fitness</h2>
<p>The ability to define custom fitness policies is perhaps the most useful feature of <b>fwdpp</b>. Broadly-speaking, there are two typical types of fitness models used in population genetics. The first are what I call site-based models, such as the model of multipicative fitness across sites. In this standard model, each non-neutral site is effectively its own gene. (A trans-heterozygotoe for two recessive mutations has wild-type fitness under the multiplicative assumption. That satisfies Benzer's definition of complementation, which is the operational definition of a gene.) The second class are models are haplotype- or region- based, in that fitness depends on the effect sizes of the maternal and paternal haplotypes that a diploid inherited. The library supports both types. Haplotype-based fitness models are often efficient to compute, but site-based models can be done badly. The library provides additional assistance for site-based models.</p>
<p><b>A fitness policy is a function or function object taking two iterators pointing to gametes are arguments are returning a double. The pointers are the diploid's haplotypes, and the return value is the fitness.</b> </p>
<h3>Site-dependent models of fitness</h3>
<p>The library provides a fitness policy called <b><a class="el" href="../../d0/ddb/structKTfwd_1_1site__dependent__fitness.html" title="Function object for fitness as a function of individual mutations in a diploid. ">KTfwd::site_dependent_fitness</a></b>. This function object requires two additional policies defining what to do with homozygous sites and heterozygous sites. These <code>homozygote'' and</code>heterozygote'' policies may be trivially defined as lambda expressions:</p>
<div class="fragment"><div class="line"><span class="keyword">struct </span>multiplicative_diploid</div>
<div class="line">{</div>
<div class="line">  <span class="keyword">typedef</span> <span class="keywordtype">double</span> result_type;</div>
<div class="line">  <span class="keyword">template</span>&lt; <span class="keyword">typename</span> iterator_type&gt;</div>
<div class="line">  <span class="keyword">inline</span> <span class="keywordtype">double</span> operator()(<span class="keyword">const</span> iterator_type &amp; g1, <span class="keyword">const</span> iterator_type &amp; g2,</div>
<div class="line">                           <span class="keyword">const</span> <span class="keywordtype">double</span> scaling = 1.)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">  </span>{</div>
<div class="line">    <span class="keyword">using</span> __mtype = <span class="keyword">typename</span> iterator_type::value_type::mutation_list_type_iterator;</div>
<div class="line">    <span class="keywordflow">return</span> site_dependent_fitness()(g1,g2,</div>
<div class="line">    <span class="comment">//Homozygote policy</span></div>
<div class="line">                                    [&amp;](<span class="keywordtype">double</span> &amp; fitness, <span class="keyword">const</span> __mtype  &amp; <a class="code" href="../../d1/dbe/crossoverTest_8cc.html#ad9baf2d3214da5c5b5710791a07c22a6">mut</a>)</div>
<div class="line">                                    {</div>
<div class="line">                                      fitness *= (1. + scaling*<a class="code" href="../../de/d89/structmut.html">mut</a>-&gt;s);</div>
<div class="line">                                    },</div>
<div class="line">                                    <span class="comment">//Heterozygote policy</span></div>
<div class="line">                                    [](<span class="keywordtype">double</span> &amp; fitness,<span class="keyword">const</span> __mtype &amp; <a class="code" href="../../d1/dbe/crossoverTest_8cc.html#ad9baf2d3214da5c5b5710791a07c22a6">mut</a>)</div>
<div class="line">                                    {</div>
<div class="line">                                      fitness *= (1. + <a class="code" href="../../de/d89/structmut.html">mut</a>-&gt;h*<a class="code" href="../../de/d89/structmut.html">mut</a>-&gt;s);</div>
<div class="line">                                    },</div>
<div class="line">                                    1.);</div>
<div class="line">  }</div>
<div class="line">};</div>
</div><!-- fragment --><p>That last variable, scaling, means that fitness is the product of <img class="formulaInl" alt="$1, 1+sh, 1+s \times \mathrm{scaling}$" src="../../form_43.png"/> over sites. This allows you to recreate results from the different parts of the literature that use <img class="formulaInl" alt="$scaling = 1$" src="../../form_44.png"/> and 2. Many classic results are based on a scaling of 2.</p>
<p>To use the multiplicative fitness policy in your simulations, this goes to <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a></b>:</p>
<div class="fragment"><div class="line">std::bind(multiplicative_diploid(),std::placeholders::_1,std::placeholders::_2,2.)</div>
</div><!-- fragment --><p>If you wish to create your own site-dependent fitness policies, the recipe is:</p>
<ol type="1">
<li>Create your own versions of the policies determining what happens in diploids homozygous vs. heterozygous for a mutation. <b>These policies are passed the current fitness from the algorithm and an iterator to a mutation object.</b></li>
<li>Write a wrapper function like multiplicative_diploid that passes these policies to __{KTfwd::site_dependent_fitness__, <em>which itself is a policy requiring that iterators to two gametes, the homozygote/heterozygote policies, and a starting value for fitness be passed to it.</em> In the example of the built-in multiplicative fitness policy, that value of <img class="formulaInl" alt="$1$" src="../../form_45.png"/> that is passed as the last argument to <b><a class="el" href="../../d0/ddb/structKTfwd_1_1site__dependent__fitness.html" title="Function object for fitness as a function of individual mutations in a diploid. ">KTfwd::site_dependent_fitness</a></b> is the initial value of fitness for a diploid. In other words, <img class="formulaInl" alt="$w = 1$" src="../../form_46.png"/> initially, and <img class="formulaInl" alt="$w$" src="../../form_47.png"/> is then modified by <b>KTfwd::multiplicative_fitness_updater_het</b> or <b>KTfwd::multiplicative_fitness_updater_hom</b> as appropriate.</li>
<li>This wrapper function is your new fitness policy.</li>
</ol>
<p>Note: the implementation <b><a class="el" href="../../d0/ddb/structKTfwd_1_1site__dependent__fitness.html" title="Function object for fitness as a function of individual mutations in a diploid. ">KTfwd::site_dependent_fitness</a></b> is a lot of iterator/pointer arithmetic.</p>
<h3>Haplotype based fitness policies</h3>
<p>Here is one that assume a user-defined mutation type with effect size <img class="formulaInl" alt="$e$" src="../../form_27.png"/> associated with it. The effect of a haplotype is additive, the genetic model is recessive, and the phenotype is the genetic effect + a Gaussian random variable with standard deviation sigma. Finally, fitnesses are under Gaussian stabilizing selection with a standard deviation of 1 and mean 0.</p>
<div class="fragment"><div class="line"><span class="keyword">struct </span>hapfitness</div>
<div class="line">{</div>
<div class="line">  <span class="keyword">typedef</span> result_type double;</div>
<div class="line">  <span class="keyword">template</span>&lt;<span class="keyword">typename</span> gam_itr</div>
<div class="line">  <span class="keywordtype">double</span> operator()(<span class="keyword">const</span> gam_itr &amp; g1,</div>
<div class="line">                    <span class="keyword">const</span> gam_itr &amp; g2,</div>
<div class="line">                    gsl_rng * r,</div>
<div class="line">                    <span class="keyword">const</span> <span class="keywordtype">double</span> &amp; sigmaE)<span class="keyword"> const</span></div>
<div class="line"><span class="keyword">   </span>{</div>
<div class="line">     <span class="keywordtype">double</span> sum1=0.,sum2=0.;</div>
<div class="line">     <span class="keyword">typedef</span> <span class="keyword">typename</span> gam_itr::value_type::mutation_container::const_iterator mci;</div>
<div class="line">     <span class="keywordflow">for</span>( mci mitr = g1.smutations.begin() ; mitr != g1.smutations.end() ; ++mitr )</div>
<div class="line">     {</div>
<div class="line">       <span class="comment">//mitr is an iterator to an iterator!</span></div>
<div class="line">       sum1 += (*mitr)-&gt;e;</div>
<div class="line">     }</div>
<div class="line">     <span class="keywordflow">for</span>( mci mitr = g2.smutations.begin() ; mitr != g2.smutations.end() ; ++mitr )</div>
<div class="line">     {</div>
<div class="line">       <span class="comment">//mitr is an iterator to an iterator!</span></div>
<div class="line">       sum2 += (*mitr)-&gt;e;</div>
<div class="line">     }</div>
<div class="line"></div>
<div class="line">     <span class="comment">//make sum1 be the value closest to 0</span></div>
<div class="line">     <span class="keywordflow">if</span>( fabs(sum1) &gt; fabs(sum2) ) { std::swap(sum1,sum2); }</div>
<div class="line"></div>
<div class="line">     <span class="comment">//add noise to fitness</span></div>
<div class="line">     <span class="comment">//Using sum1 as the genetic part makes the model recessive</span></div>
<div class="line">     <span class="keywordtype">double</span> pheno = sum1 + gsl_ran_gaussian(r,sigmaE);</div>
<div class="line"></div>
<div class="line">     <span class="comment">/*</span></div>
<div class="line"><span class="comment">       Return fitnees under Gaussian stabilizing model.</span></div>
<div class="line"><span class="comment">       This is the only part of the Gaussian pdf that</span></div>
<div class="line"><span class="comment">       matters.  The rest is a constant</span></div>
<div class="line"><span class="comment">       and so won&#39;t affect sampling prop to fitness.</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">     return std::exp( -std::pow(pheno,2) /2. );</span></div>
<div class="line"><span class="comment">   }</span></div>
<div class="line"><span class="comment">};</span></div>
</div><!-- fragment --><p>To use the above policy:</p>
<div class="fragment"><div class="line">std::bind( hapfitness(),</div>
<div class="line">std::placeholders::_1,std::placeholders::_2,</div>
<div class="line">r,sigmaE );</div>
</div><!-- fragment --><p>Please note that a long-running annoyance with open-source C++ compilers (GCC!) is whether or not exp, pow, log, etc., are in namespace std or in the global namespace. This can vary from version to version and across operating systems. Sometimes, you need to say <b>std::exp</b> when on another system that fails and you need __::exp__.</p>
<h2><a class="anchor" id="TutUpdateRemove"></a>
Updating and removal policies</h2>
<p>During the course of a simulation, new mutation and gamete types come and go. New types must be entered in to their approproate containers. The relevant policies are basically searches followed by either a member variable update or an insertion. For example, consider our mutation model policy in section infsites}. The mutation returned from that function is not currently found in our mutation list (because it has a unique position). Therefore, the relevant policy is just to insert the new mutation at the end of the doubly-linked list of mutations. Likewise, a gamete with a new mutation from that model cannot be identical to any currently-segregating gamete. Therefore, a good policy is just to insert it at the end of the doubly-linked list of gametes (or push it back to the vector of gametes in a gamete-based simulation). However, a recombination event could make either a new gamete or a pre-existing gamete. Thus, a policy to insert a recombinant gamete should do the following:</p>
<ol type="1">
<li>Figure out of the gamete is new or not.</li>
<li>If it is new, add it to the end of the gamete container.</li>
<li>It it is not new, do something intelligent, such as return the iterator to the pre-existing version of that gamete in the container.</li>
</ol>
<p>These types of policies are simple. See the header files <b><a class="el" href="../../d0/d13/insertion__policies_8hpp.html">fwdpp/insertion_policies.hpp</a></b> and <b><a class="el" href="../../d4/d62/fwd__functional_8hpp.html">fwdpp/fwd_functional.hpp</a></b>. The latter contains the functions passed to <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a86a4fcecb15c101816ccdd15eefd6a1b" title="Remove mutations from population Removes mutations that are lost. ">KTfwd::remove_lost</a></b> and <b><a class="el" href="../../da/d8b/namespaceKTfwd.html#a84e666ea27d38bdfc539ffd23653eaa9" title="Remove mutations from population Removes mutations that are fixed or lost. ">KTfwd::remove_fixed_lost</a></b>&ndash;see the example programs (and the next section here) for concrete examples.</p>
<h1><a class="anchor" id="TutTieItup"></a>
Putting it all togeter (kinda)</h1>
<p>Ok, once we have defined our mutation types, our containers, and our policies, a single generation of a constant-sized population in an individual-based simulation is evolved a follows:</p>
<div class="fragment"><div class="line"><span class="keyword">using</span> std::placeholders; <span class="comment">//_1,_2, etc.</span></div>
<div class="line"><span class="keywordtype">double</span> wbar = <a class="code" href="../../da/d8b/namespaceKTfwd.html#a946efaa4a6095f4f2aadda44a809a477">KTfwd::sample_diploid</a>(r,</div>
<div class="line">                                    &amp;gametes,  <span class="comment">//non-const pointer to gametes</span></div>
<div class="line">                                    &amp;diploids, <span class="comment">//non-const pointer to diploids</span></div>
<div class="line">                                    &amp;mutations, <span class="comment">//non-const pointer to mutations</span></div>
<div class="line">                                    N,     <span class="comment">//current pop size, remains constant</span></div>
<div class="line">                                    mu,    <span class="comment">//mutation rate per gamete</span></div>
<div class="line">                                    <span class="comment">/*</span></div>
<div class="line"><span class="comment">                                    The mutation model (defined above) will pass each gamete</span></div>
<div class="line"><span class="comment">                                       to be mutated to the mutation model function.  Again, _1</span></div>
<div class="line"><span class="comment">                                       is used as a placeholder for that gamete.</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">                                     std::bind(mutmodel_Qtrait,r,_1,mu_neutral,mu_selected,sigma_e,&amp;lookup),</span></div>
<div class="line"><span class="comment">                                     //The recombination policy includes the recombination map policy</span></div>
<div class="line"><span class="comment">                                     std::bind(KTfwd::genetics101(),_1,_2,</span></div>
<div class="line"><span class="comment">                                                 &amp;gametes,</span></div>
<div class="line"><span class="comment">                                                 littler,</span></div>
<div class="line"><span class="comment">                                                 r,</span></div>
<div class="line"><span class="comment">                                                 recmap),</span></div>
<div class="line"><span class="comment">                                     /*</span></div>
<div class="line"><span class="comment">                                       Policy to insert new mutations at the end of the mutations list</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">                                     std::bind(KTfwd::insert_at_end&lt;mtype,mlist&gt;,_1,_2),</span></div>
<div class="line"><span class="comment">                                     /*</span></div>
<div class="line"><span class="comment">                                       Policy telling KTfwd::mutate how to add mutated gametes into the gamete pool.</span></div>
<div class="line"><span class="comment">                                       If mutation results in a new gamete, add that gamete to the </span></div>
<div class="line"><span class="comment">                                       end of gametes. This is always the case under infinitely-many sites,</span></div>
<div class="line"><span class="comment">                                       but for other mutation models, mutation may result in a new</span></div>
<div class="line"><span class="comment">                                       copy identical to an existing gamete.  If so,</span></div>
<div class="line"><span class="comment">                                       that gamete&#39;s frequency increases by 1.</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">                                     std::bind(KTfwd::insert_at_end&lt;gtype,glist&gt;,_1,_2),</span></div>
<div class="line"><span class="comment">                                     std::bind(  std::bind( hapfitness(),</span></div>
<div class="line"><span class="comment">                                                 _1,_2,</span></div>
<div class="line"><span class="comment">                                                 r,sigmaE ),</span></div>
<div class="line"><span class="comment">                                     /*</span></div>
<div class="line"><span class="comment">                                       Only remove lost mutations</span></div>
<div class="line"><span class="comment"></span></div>
<div class="line"><span class="comment">                                     std::bind(KTfwd::mutation_remover(),_1,0));</span></div>
<div class="line"><span class="comment">        //Clean up the mutations list.  This also resets ``checked&#39;&#39; in each mutation to zero,</span></div>
<div class="line"><span class="comment">        //which is that ``internal detail&#39;&#39; referred to above</span></div>
<div class="line"><span class="comment">        //We don&#39;t removed fixed mutations b/c this is a quantitative trait simulation</span></div>
<div class="line"><span class="comment">        KTfwd::remove_lost(&amp;mutations,&amp;lookup,generation);</span></div>
</div><!-- fragment --> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="../../doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
